src/world_interface/base_motor.h

```diff
@@ -0,0 +1,144 @@
+#include "../utils/scheduler.h"
+
+using namespace std::chrono_literals;
+
+/**
+ * @namespace robot
+ * @brief Collection of functions for manipulating a motor.
+ */
+namespace robot {
+/**
+ * An abstract motor class
+ * 
+ */
+class base_motor {
+public:
+
+    // TODO: spin up thread in constructor, use it for setMotorVel (static mutex lock)
+
+    /**
+     * @brief Default constructor for base motor.
+     *
+     */
+    base_motor();
+
+    /**
+     * @brief Returns the status of the motor position sensor.
+     *
+     * @return True if the motor has a position sensor and false if not
+     */
+    virtual bool hasPositionSensor() const = 0;
+
+    /**
+     * @brief Set the PWM command of the motor.
+     *
+     * @param power The power command, in the range [-1, 1]
+     */
+    virtual void setMotorPower(double power) = 0;
+
+    /**
+     * @brief Set the target position of the motor. This will have no effect if the motor
+     * does not support PID.
+     *
+     * @param targetPos The target position, in millidegrees. Refer to the specific motor for more
+     * information.
+     */
+    virtual void setMotorPos(int32_t targetPos) = 0;
+
+    /**
+     * @brief Get the last reported position of the specified motor.
+     *
+     * @return types::DataPoint<int32_t> The last reported position of the motor, if it exists.
+     * If the motor has not reported a position (because it hasn't been received yet or if it
+     * doesn't have an encoder) then an empty data point is returned.
+     */
+    virtual types::DataPoint<int32_t> getMotorPos() const = 0;
+
+    /**
+     * @brief Sets the velocity of the motor.
+     *
+     * @param targetVel The target velocity, in millidegrees per second.
+     */
+    void setMotorVel(int32_t targetVel) {
+        // create a velocity controller if it's not already initialized
+        if (!velController) {
+            constructVelController();
+        }
+        assert(("Vel controller is initialized", velController));
+
+        // set velocity target
+        navtypes::Vectord<1> velocityVector {targetVel};
+        types::datatime_t currTime = types::dataclock::now();
+        velController->setTarget(currTime, velocityVector);
+        assert(("Target velocity should be set in setMotorVel", velController->hasTarget()));
+
+        // check to see if the event exists. if yes, unschedule it
+        resetEventID();
+
+        // schedule position event
+        velEventID = pSched.scheduleEvent(100ms, [&]() -> void {
+                        types::datatime_t currTime = types::dataclock::now();
+                        const navtypes::Vectord<1> currPos {getMotorPos().getData()};
+                        navtypes::Vectord<1> posCommand = velController->getCommand(currTime, currPos);
+                        // TODO: is there only one position value in the matrix?
+                        setMotorPos(posCommand.coeff(0, 0));
+                    });
+    }
```

Comment:
Can we put this implementation in a `base_motor.cpp`? Doesn't need to be in the header file, it seems.

Reply:
Done